Proxy infrastructure for NodePorts

justinsb opened the pull request:
A service with a NodePort set will listen on that port, on every node.

This is both handy for some load balancers (AWS ELB) and for people
that want to expose a service without using a load balancer.

Part of the Visibility PR stream...

justinsb committed:
Proxy infrastructure for NodePorts

A service with a NodePort set will listen on that port, on every node.

This is both handy for some load balancers (AWS ELB) and for people
that want to expose a service without using a load balancer.

Part of the Visibility PR stream...

justinsb commented:
@thockin This is the changes to proxier for NodePorts pulled out, as I think you wanted to have a look.  Hope it is easier to see now that it is just this patch!

(The intent of each of these PRs is that - even if they don't actually do anything individually - they should all be mergeable and the final PR activates them all!)

thockin commented on pkg/proxy/proxier.go:
```diff
@@ -328,6 +332,9 @@ func (proxier *Proxier) openPortal(service ServicePortName, info *serviceInfo) e
 			return err
 		}
 	}
+	if info.nodePort != 0 {
```
We should keep a map[int]bool of all the ports (nodePorts and proxy ports, maybe) we have allocated and fail if there is a conflict.

justinsb replied:
(This comment was on an old version, I believe, and this is done now: portMap)

thockin commented on pkg/proxy/proxier.go:
```diff
@@ -404,6 +466,14 @@ func iptablesInit(ipt iptables.Interface) error {
+	// TODO (comment from thockin): This is a little scary - if somehow these rules get out of order, this will consume
+	// all accesses to a given port number regardless of destIP.
+	// Should this maybe use -m addrtype --dst-type LOCAL before -j ?
+	// And I think this should be installed in the OUTPUT chain too, no? PREROUTING is from off-machine or from
```
You can nix this comment from here down.

thockin commented on pkg/proxy/proxier.go:
```diff
@@ -404,6 +466,14 @@ func iptablesInit(ipt iptables.Interface) error {
 		return err
 	}
+
+	// TODO (comment from thockin): This is a little scary - if somehow these rules get out of order, this will consume
```
I stand by this comment - I want to play a bit more to come up with a suggested implementation.  Will have time later tonight.

thockin replied:
I played.

`-m addrtype --dst-type LOCAL` is what we want *except* that it is broken (by intent without foresight to our usecase) on at least GCE.  Specifically, GCE machines have a daemon which learns what external IPs are forwarded to that machine, and configure a local route for that IP, making a match for  `--dst-type LOCAL` when we don't want it to.

Removing the route gives correct behavior until the daemon recreates it.

Killing the daemon is an option, but means that any non-kubernetes use of the machine with external IP will be broken.

I think the simplest thing we can do for now is to:

1) leave a TODO to enable `-m addrtype --dst-type LOCAL` with the above comments about GCE
2) make sure this jump happens last, comment it carefully
3) add a `-m comment --comment "handle service NodePorts; NOTE: this must be the last rule in the chain"` to this jump rule

What think?

justinsb replied:
One option would be to jump from chain 1 to chain 2, instead of doing -j chain1 / return / -j chain 2

```
-A INPUT -j CHAIN1

-A CHAIN1 -j CHAIN2
```

instead of

```
-A INPUT -j CHAIN1
-A INPUT -j CHAIN2
```

The downside is that then we must be careful that we already insert rules into CHAIN1 ahead of the jump to CHAIN2.

All-in, I think I prefer keeping the logic simpler and doing what you suggested in your note.  We could add a test that the jumps are in the correct order in kube-proxy.  And I think this will probably get revisited with the work to move to pure iptables.

thockin replied:
Make it so.

I considered tail-jumping, but iptables individual commands don't do a good
job of making this easy.  Once we switch to a full iptables-restore sort of
mode (atomic) then we can ensure ordering.

On Mon, May 18, 2015 at 9:50 AM, Justin Santa Barbara <
notifications@github.com> wrote:

> In pkg/proxy/proxier.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/8383#discussion_r30525573>
> :
>
> > @@ -404,6 +466,14 @@ func iptablesInit(ipt iptables.Interface) error {
> >  	if _, err := ipt.EnsureChain(iptables.TableNAT, iptablesContainerPortalChain); err != nil {
> >  		return err
> >  	}
> > +
> > +	// TODO (comment from thockin): This is a little scary - if somehow these rules get out of order, this will consume
>
> One option would be to jump from chain 1 to chain 2, instead of doing -j
> chain1 / return / -j chain 2
>
> -A INPUT -j CHAIN1
>
> -A CHAIN1 -j CHAIN2
>
> instead of
>
> -A INPUT -j CHAIN1
> -A INPUT -j CHAIN2
>
> The downside is that then we must be careful that we already insert rules
> into CHAIN1 ahead of the jump to CHAIN2.
>
> All-in, I think I prefer keeping the logic simpler and doing what you
> suggested in your note. We could add a test that the jumps are in the
> correct order in kube-proxy. And I think this will probably get revisited
> with the work to move to pure iptables.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/8383/files#r30525573>
> .
>

thockin commented on pkg/proxy/proxier.go:
```diff
@@ -549,6 +641,37 @@ func (proxier *Proxier) iptablesHostPortalArgs(destIP net.IP, destPort int, prot
 
+// Build a slice of iptables args for a from-container public-port rule.
+// See iptablesContainerPortalArgs
+// TODO: Should we just reuse iptablesContainerPortalArgs?
```
I'm inclined to say yes, but don't change that iuntil I have a concrete proposal for the addrtype idea above.

thockin replied:
leave this as a TODO

thockin commented:
LGTM overall, I'll come back with a final LGTM ASAP.

justinsb committed:
Detect nodePort conflicts in proxier

We only check for nodePort/nodePort conflicts right now, not e.g. nodePort/portalPort conflicts

justinsb committed:
Filter NodePort rules on addrtype=LOCAL

And add some notes about potential complications

justinsb commented:
@thockin Going to put this through testing, but wanted to give you an early look...

justinsb committed:
Insert rules that should be first, append rules that should be last

thockin commented on pkg/proxy/proxier.go:
```diff
@@ -368,12 +381,80 @@ func (proxier *Proxier) openOnePortal(portalIP net.IP, portalPort int, protocol
+		proxier.portMap[port] = owner
+		return nil
+	}
+	if existing == owner {
```
How would this ever happen?  This would indicate a pretty serious bug, I think.  I'd rather not silently succeed.

justinsb replied:
We sync the rules periodically.

SyncLoop -> ensurePortals -> openPortal -> openNodePort -> claimPort

thockin commented on pkg/proxy/proxier.go:
```diff
@@ -368,12 +381,80 @@ func (proxier *Proxier) openOnePortal(portalIP net.IP, portalPort int, protocol
 	return nil
 }
 
+func (proxier *Proxier) lockPort(port int, owner ServicePortName) error {
```
I don't think lock/unlock is the right analogy.  Claim/release.  Take/Put.  Something like that

thockin commented on pkg/proxy/proxier.go:
```diff
@@ -464,9 +617,13 @@ func iptablesCommonPortalArgs(destIP net.IP, destPort int, protocol api.Protocol
 		"--dport", fmt.Sprintf("%d", destPort),
 	}
+
+	if destIP != nil {
```
rather than passing nil and testing it, why not just move -d into the portal-specific functions that call this function.

justinsb replied:
I feel this code is pretty fragile, and it is nice to contain it in iptablesCommonPortalArgs.  So I'd lobby against doing this, but if you do really want it then I'll make the change.

thockin replied:
That's fine.  I just want to find a better way to factor all of these that is more self-describing.

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -28,6 +28,13 @@ import (
+type RulePosition string
+
+const (
+	RulePositionFirst RulePosition = "-I"
```
call these Prepend and Append, I think it will read easier.

thockin commented:
Looks plausible.

justinsb committed:
Fix-up portMap

justinsb committed:
Rename RulePositionFirst/Last => Prepend/Append

thockin commented:
LGTM

Do not commit until the full load-balancer PR sequence is complete

justinsb commented:
NodePort is now merged, in a parallel PR
